### ISSUE TITLE
No equivalent to LEAST and GREATEST in SQL ?

### ISSUE THREAD

**Issue by maelp** [PRQL/prql#4151]
### What's up?

I have a table where each row has for example

temperature_a, temperature_b, etc

and I want to create a condition if min(temperature_a, temperature_b) < somethreshold OR max(temperature_a, temperature_b) > someOtherThreshold

in SQL I can do something like

```
    SELECT
        LEAST(value.temperature_front, value.temperature_back) as min_temp_cells,
        GREATEST(value.temperature_front, value.temperature_back) as max_temp_cells,
        GREATEST(value.temperature_mosfet, value.temperature_ic) as max_temp_bms
   FROM ...
```

but I don't see the equivalent in PRQL?

**Comment by maelp** [comment]
What would be a nicer way to do this than

```
from t = s"SELECT * FROM recent_cse(4)"
derive {
  min_temp_cells = s"LEAST(value.temperature_front, value.temperature_back)",
  max_temp_cells = s"GREATEST(value.temperature_front, value.temperature_back)",
  max_temp_bms = s"GREATEST(value.temperature_mosfet, value.temperature_ic)",
}
filter min_temp_cells < -10 || max_temp_cells > 40 || max_temp_bms > 70
group bms_id (
  aggregate {
    min_temp_cells = min min_temp_cells,
    max_temp_cells = max max_temp_cells,
    max_temp_bms = max max_temp_bms,
    count = count this,
  }
)
```

**Comment by max-sixty** [comment]
Yes we should probably add these functions. Using s-strings is the best approach for the moment.

---

For adding them — do we want to use `LEAST` & `GREATEST`? I always found these were slightly awkwardly named — they seem to use synonyms for `MIN` & `MAX` to to distinguish them from aggregation functions. But the words don't indicate whether they're aggregation functions or not.

We could instead use `math.min` & `math.max` given we have the advantage of modules. But possibly that's confusing...

**Comment by maelp** [comment]
I would probably keep least and greatest for consistency, or perhaps just have an "overload" of min which can be used when there are multiple params, eg `min(someField)` would be the aggregate of someField over the rows, and `min(fieldA, fieldB, fieldC)` would be the min of the field for each row (but might be confusing, don't know if there are contexts in which they could have both an "in-row" and "across row" meaning?)

**Comment by parsons-kyle-89** [comment]
Hi, I thought I might work on this as it seemed like an easy first issue.  Still getting my bearings on everything.  It seems like a lot of functions are defined as s-strings in `prqlc/prqlc/src/sql/std.sql.prql`.  It looks like the supported dialects are Clickhouse, DuckDB, MariaDB, MySQL, Postgres, SQLite, MSSQL (T-SQL?), BigQuery, and Snowflake.  Going through the docs of all of those helpfully linked in that file, it seems like they _all_ have `GREATEST` and `LEAST` except for SQLite which does the overloaded `MAX`/`MIN` idea above.

So to add this (basically just following my nose and comparing to something like `rtrim`):
1) Add a definition at the top level of `src/sql/std.sql.prql` and then also provide an overload definition in the SQLite module?
2) Add further references in `src/semantic/std.prql`
3) Add some tests
4) Add to the grammar files
5) Add to the documentation

What's not clear to me is how to support an n-ary version of `GREATEST` and `LEAST`.  Looking at analogues, it seems there's only binary `COALESCE` and that `CASE` is a special construct.  Is it as simple as providing more overloads with more arguments?  (Though that sees like a somewhat limited approach.)

Edit:  I tried adding an overload to `GREATEST` with more arguments and it didn't seem to work.  It raised the error `Error: duplicate declarations of std.greatest` which seems pretty clear.

**Comment by NickCrews** [comment]
+1 to this overloading syntax:

> min(someField) would be the aggregate of someField over the rows, and min(fieldA, fieldB, fieldC) would be the min of the field for each row

I would prefer that over a separate GREATEST and LEAST function.

Note, for prior art, [ibis went with greatest and least](https://ibis-project.org/reference/expression-generic#ibis.least).
